### ISSUE TITLE
Question about media platform.

### ISSUE THREAD

**Issue by xannor** [starkillerOG/reolink_aio#9]
Do you have any plans to build out a media platform using the search command? If you dont, or dont plan on doing it for a while, I might create a custom integration that wraps around the "built in" reolink and this api to implement one. I had built the original one in the fwesterberg custom integration, and has plans to build one in my own CI until your built in one surpassed it faster than I could work on it.

My idea is the media platform providing date sorted access to recordings and a custom sensor that provides info on the last recording for a channel. The only real drawback is that the rest api does not provide any trigger information other than a date/time, (but the client one does) and no potential thumbnails.

any thoughts?

**Comment by starkillerOG** [comment]
@xannor thank you very much for reaching out and your contributions to the projects on which this library is based on!

I was not planning on adding the media platform on the short term, but I would be more than happy to review and accept PRs (both in this library and in HomeAssistant) if you are willing to work on it.
I am more than happy to help out if you get stuck somewhere.

Your idea sounds good and was indeed already pretty much incoporated in the  fwesterberg custom integration.
However there was a lot of complicated code to locally store snapshots that were used as thumbnails for the media platform.
I don't think that was a good idea, it added a lot of complexity to the code and it took a lot of storage drive space on the HomeAssistant device which then needed cleanup services etc. etc.
Bottom line: I would suggest not adding thumbnails unless we can convince reolink to build that into their firmware such that we can simply request a thumbnail for a video from the search command (and preferably also info on the trigger for that video).

**Comment by xannor** [comment]
I agree, I did not like it either, but "live" thumbs on a raspberry at the time were much worse, as FFMPEG was incredibly slow. But good to know, I'll first tool around with an "extension" integration to start with, and see if I can piggyback off the the official. Then when things are stable we can see about PR's to push parts over.

**Comment by starkillerOG** [comment]
@xannor good to hear you have started to work on it!
To answer your questions:
1) The channel names are exposed by the `camera_name` property: https://github.com/starkillerOG/reolink_aio/blob/2362257923badb708543643a2f1e3c7bcb07cdb5/reolink_aio/api.py#L399
The DUO 1st gen might be an special case not handled properly yet, it is a dual lens camera, I do not have one myself and I don't now how it will be handled, I believe it just has two channels and schould therefore also have two channel names, but I have not tested this.

2) There is a `_subscription_time_difference` internall property in the library which gives the time diffrence in seconds between the UTC time of the homeassistant device and the `CurrentTime` on the camera.
I will investigate what this "CurrentTime' is, I think it is actually also UTC.
I don't quite get why you would need this, but if you need it, just add it in a PR.

**Comment by xannor** [comment]
1) didnt see that, good to know.

2) The reason I wanted the offset is because the times for recordings are in camera time, with no tz info so to make sure they are accurate I need to convert them to UTC for HA. CurrentTime was just my reference to the GetTime call (I couldnt remember the name when I posted.) The only "reliable" way to convert them is to have the full time info, which includes the timezone, since I have to deal with historical datetimes that can cross daylight savings time zones. I can use the private value you hold with that info. 

As for the DUO, it reports channels like an NVR, if you need and debug dumps from it I can provide them, and possibly answer some questions about it as that is what I primarily tested and developed against.

**Comment by starkillerOG** [comment]
@xannor did you setup the DUO 1st gen in the new build in HomeAssistant integration?
Did you experiance any issues or does all look good?
Could you share screenshots of the device pages of both channels?
What was the initial naming of the two devices, and did it also make a third device (NVR)?

**Comment by starkillerOG** [comment]
@xannor about the time, if you now the time_diffrence in seconds between the utc.now() on HomeAssistant and the camera internal time you can always calculate the time you need. Just take the datetime from the recording and add/substract the time diffrence in seconds, to get UTC on the HomeAssistant device.

Normally this time diffrence will only be a second or so, but if the camera timezone is set incorrectly lets say 2 hours in front the timediffrence will become 7200 seconds (two hours).

But I think the conversion/correction to UTC timezone schould happen inside the reolink-aio library in the `request_vod_files` function if that is not yet done.

I have not yet looked at the `request_vod_files` function, but that needs to be cleaned up and probably some more parsing/processing of the resulting data schould happen in that function.
Probably that function schould return a list of video files available with each item something like a dataclass wich has all the info we need and already processed such that it can easily be used in HomeAssistant.

**Comment by starkillerOG** [comment]
@xannor in this commit I have already cleaned up the request_vod_files method a bit https://github.com/starkillerOG/reolink_aio/commit/4fa3a87ecb71be99b50c38ec9d250e7d9eb0ea2e

But further processing is probably a good idea, convert the dict to a class with date_time objects with all time conversions already taken care of

**Comment by starkillerOG** [comment]
Also cleaned up get_vod_source method a bit in this commit: https://github.com/starkillerOG/reolink_aio/commit/80affcd0abb6a6ea3ee921ed861503d48d9f6b23

**Comment by xannor** [comment]
Time:
The main concern/issue is that the camera does not store or present in UTC but localtime w/o tz info. For a non DST timezone this is no issue, however for a TZ that uses DST half the year, historical times would be an hour off. So if you have a device with recordings that cross a time change, without the correct TZ info, they would be off. I have code that I wrote in my libs that would parse the GetTime/Dst info into an actual python tzinfo, and when merged with the Time data would give a datetime with the correct TZ that can be used to convert to/from UTC/localtime for any date. [TZ class](https://github.com/xannor/py_reolinkapi/blob/main/src/async_reolink/api/commands/system.py#L44)

DUO:
I installed it when I started working on the media stuff about a week ago, right now into, 2022.3.4. I had no issues with the setup, and it did create entities for both cameras in the device, but as I stated, both sets of entities were created with the same names, as if two separate cameras were mashed together, instead of being prefixed with the name of the channel. Also only 1 device was created, which is ok, because in my code I had created a main device and a sub device for each channel, but it always looked/felt weird to me that HA would list all three devices, and not duplicate missing info into the sub devices, so they always looked incomplete to me.

**Comment by xannor** [comment]
I have found something to add though, for channel capabilities, recDownload and recReplay. These should show if a channel is allowed to use the Download action or if it needs to replay via url (which I think may be less optimal in most cases.) I see you have code to get the vod_source, but not to download the file direct (3.6.6 in the api doc) I know the browser interface does the download call, and I think we can "stream" the download.

**Comment by xannor** [comment]
I started going down the path of using a Calendar as the "entity" managing the search (it mostly fist concept wise.) and in the process I re-build my tz code from my other integration. If you are curios about and/or want to critique it: [DateTime utils](https://github.com/xannor/ha_reolink_extras/blob/dev/custom_components/reolink_extras/util/dt.py)

Seems to work so far but it is still very much experimental.

**Comment by starkillerOG** [comment]
> I have found something to add though, for channel capabilities, recDownload and recReplay. These should show if a channel is allowed to use the Download action or if it needs to replay via url (which I think may be less optimal in most cases.) I see you have code to get the vod_source, but not to download the file direct (3.6.6 in the api doc) I know the browser interface does the download call, and I think we can "stream" the download. This may work around the rtmp HD issue as well.

If we use the download option, where does the file get downloaded to?
Is it automatically removed from the HomeAssistant device when we stop streaming from HomeAssistant?

**Comment by xannor** [comment]
From the API Docks the download action just provides the MP4 as a data stream over http as apposed to pushing through one of the rtmp/rtsp servers. Since the the files are complete I figure this would be better on the camera.

**Comment by starkillerOG** [comment]
> From the API Docks the download action just provides the MP4 as a data stream over http as apposed to pushing through one of the rtmp/rtsp servers. Since the the files are complete I figure this would be better on the camera.

Have you tried this with HomeAssistant?
Is HomeAssistant capable of handeling MP4 data streams?

I am starting to think it would be wise to have a configuration option to specify the `playback protocol` with options such as MP4, RTSP, RTMP etc.....

**Comment by xannor** [comment]
I have not tried the download command from the API yet, but I have used the replay through RTMP, and I am hoping the download method will solve some of the issues I found.

1) Last I checked, only RTMP supports playback of recorded, but RTMP cannot send 4k so only sub-streams or less could be transferred.
2) RTMP playback is every bit as slow to launch as the live stream.
3) Either the RTMP service, or the HA player, expects the stream to be live, so you lose any ability to seek the stream and do not know its length, only what has buffered so far. Also when the file ends the stream just freezes, nothing indicates and end of stream.

I dont plan on actually downloading full files to the HA install, just streaming the download through the camera stream service. I am hoping this would both load faster and provide the ability to seek the playback, in addition to providing the full quality, and possibly the ability of the user to save the playback somewhere if they choose.

**Comment by xannor** [comment]
OffTopic>
I just updated my test environment to HA 23-4-2 and the DUO does show up a little better there. One thing I would like to see is the default names for the channel based sensors/entries to inherit the channel name of the lens, right now only the cameras inherit the name so I get two motion sensors with the same name for each type of sensor. Also, and this is something I dont know how to fix, even though the duo allows and presents the floodlight on each channel, only 1 switch physically exists so flipping either will turn the light on. This is not an issue except that the other switch does not update when you flip them so it looks odd. No a major issue but something you might get complaints about if someone does not know the difference.
/OffTopic

On the topic front, I am learning to "Decode" the vod filenames, I believe they have the motion sensor data encoded in them, and I can use that in the search to "client side" filter and display the triggers for the file. Unfortunatly I can only get filenames for pet/person/motion/and timer, I dont have a camera in a location that can do vehicle detection, and I dont have anything that supports the other types of detection in the API (face, animal, etc). I think the format has space for those types but I cannot verify.

**Comment by xannor** [comment]
ran into a snag with your library. You code assumes that if a VOD search returns no files but files were requested, it is an error.  I think the cameras (at least the duo I am testing with) will also return no files and just a status if there are too many results. While technically correct, it also eats the Status entry so I cannot trap and work around the error (i.e. know which days of the month do contain files and split my request up). If you could either not throw when no file results and pass an empty list, or throw a custom error with the status attached that I can trap, that would help "Work around" this issue.

**Comment by starkillerOG** [comment]
@xannor thanks for figuring out this bug, my NVR always returns a (very long) list of recordings, but if I chose a time-intervall where there are no recordings I would indeed get the error as you posted. I have fixed by simply return a empty list. 

Besides I already did some of the parsing in this library using the new VOD_file class. Please have a look at it, and add properties if you need more in a PR.
It is available in the just released reolink_aio v0.5.13

My simple test code:
```
vod_files = await host.request_vod_files(channel=0, start=datetime.now() - timedelta(days=2), end=datetime.now(), stream="sub")
for file in vod_files[1]:
        print(file)
        print(file.file_name)
```

If your DUO cam does not return any results if there schould be results because there are too many, please make some additonal code that re-requests the vod_files in chunks and stitches them together afterwords.

**Comment by starkillerOG** [comment]
@xannor 

> I am learning to "Decode" the vod filenames

Please add this decoding to the new VOD_file class such that we have a simple property (preferably a ENUM) that indicates the type of the recording, we can default to unkown if not available.

**Comment by xannor** [comment]
I looked over your code, and even though you mean well, your simplified UTC code has issues. Mostly, it does not handle daylight savings time, so a camera in a DST zone would be an hour off half the year if the VODs looked up were recorded in before/after a time change. This is because the camera always stores and reports in local time. This is why I went through the trouble of parsing the DST field so I can use the actual TZ info of the camera along with the date/time of the file to determine the actual local time of the file.

For the "flags" because a video can be flagged as multiple "triggers" I am looking into if I should use a intflag or an array of enums, a VOD could have multiple, even potentially none.

defaulting to an empty file list is fine, though I think there is a meaning difference in the API between no files and a missing files. I have noticed that when my search range exceeds the camera history, it would just report the statuses (i.e. as a hint of how far back you can search) even when the range would include files.

**Comment by starkillerOG** [comment]
> I looked over your code, and even though you mean well, your simplified UTC code has issues. Mostly, it does not handle daylight savings time, so a camera in a DST zone would be an hour off half the year if the VODs looked up were recorded in before/after a time change. This is because the camera always stores and reports in local time. This is why I went through the trouble of parsing the DST field so I can use the actual TZ info of the camera along with the date/time of the file to determine the actual local time of the file.

I did check for daylight savings time, and it apear to work, the `cam_hour_offset = round(self.time_offset / 3600) ` schould take care of that, but maybe I am wrong.
Although I do image indeed around the time change things would go wrong. So when the DST has just changed and I am looking at recordings that have been recordered before the DST. Does your PR https://github.com/starkillerOG/reolink_aio/pull/23 take care of this correctly (my implementation does not)?
However for my NVR I did notice that the reported start time of a recording from the Search result changes when changing the DST time setting, so the Search result is not static but also changes the past files when a DST time change happens, so I was guessing my implementation would actually work just fine.

> For the "flags" because a video can be flagged as multiple "triggers" I am looking into if I should use a intflag or an array of enums, a VOD could have multiple, even potentially none.

Ah right, did not consider that.
Maybe then just add a bunch of boolean properties that indicate if a recording has motion/people/vehicle etc event

> defaulting to an empty file list is fine, though I think there is a meaning difference in the API between no files and a missing files. I have noticed that when my search range exceeds the camera history, it would just report the statuses (i.e. as a hint of how far back you can search) even when the range would include files.

I never got back a empty file list from my NVR, did you?

**Comment by starkillerOG** [comment]
> Some feedback on your VOD_file class, What is [playback time](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L74)? I have not seen that field on my camera, I assume it is an NVR thing (I dont have one of those) 

Actually I am not entirely sure yet, it is another timestamp, I think a motion recording (short recording 1 min or so) will have the playback time refering to the start of the bigger 1 hour recording file of which it is part, or something simular, but have not figured out what it actually means. It was just in the return data from the NVR.

> Pulling that field, when it is not present would error. 

We schould default it to None when it is not present, can be done in a followup PR.

Also, why are you taking a [dict[str,any]](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L34) instead of using your [SearchFile](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L16) typeddict?

Yea that schould probably indeed be the SearchFile typing, but I just wanted to provide some basics to you of what I had in mind. I did not check if the SearchFile typing is consistant with the NVR response.

**Comment by xannor** [comment]
> I did check for daylight savings time, and it apear to work, the `cam_hour_offset = round(self.time_offset / 3600) ` schould take care of that, but maybe I am wrong. Although I do image indeed around the time change things would go wrong. So when the DST has just changed and I am looking at recordings that have been recordered before the DST. Does your PR #23 take care of this correctly (my implementation does not)? However for my NVR I did notice that the reported start time of a recording from the Search result changes when changing the DST time setting, so the Search result is not static but also changes the past files when a DST time change happens, so I was guessing my implementation would actually work just fine.

Be carefull relying on "it works on the NVR" it is better  hardware than the cameras (and tends to have more bells an whistles) so it may do things  the cameras cant. I.E. store in UTC and adjust to local (I dont own an NVR so I dont know what it is fully capable of, I can only guess/assume). The cameras I have tested with only seem to store in local time so changing the time or DST enabled does not seem to affect existing recordings, only new ones. The code I provided just implements their DST rules as a python timezone, with that you can convert the camera localtime to any time. 
> 
> > For the "flags" because a video can be flagged as multiple "triggers" I am looking into if I should use a intflag or an array of enums, a VOD could have multiple, even potentially none.
> 
> Ah right, did not consider that. Maybe then just add a bunch of boolean properties that indicate if a recording has motion/people/vehicle etc event

I would prefer to at least use enums in a list as I dont have enough sources to test and determine what all the flags values could be, and there could be more in the future. So far I can verify Timer Motion Pet and Person. I have a couple of cameras that support vehicle, but are not in a location that has any to detect. I know in the API there is also animal (maybe the Trax does this) and face. The only issue with intflags is, depending in the python version. 0 could mean none or all, and I think it may be possible to get a recording with no triggers, not sure how but I have gotten ones with pet and person but no motion.

> > defaulting to an empty file list is fine, though I think there is a meaning difference in the API between no files and a missing files. I have noticed that when my search range exceeds the camera history, it would just report the statuses (i.e. as a hint of how far back you can search) even when the range would include files.
> 
> I never got back a empty file list from my NVR, did you?

I have seen it in the past but that may have been older firmware. I will have to test and see what I get for a response in a range with no recordings, but within the status window. Not that it matters, as long as I get the status I can see if there would have been matches.

**Comment by xannor** [comment]
> > Some feedback on your VOD_file class, What is [playback time](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L74)? I have not seen that field on my camera, I assume it is an NVR thing (I dont have one of those)
> 
> Actually I am not entirely sure yet, it is another timestamp, I think a motion recording (short recording 1 min or so) will have the playback time refering to the start of the bigger 1 hour recording file of which it is part, or something simular, but have not figured out what it actually means. It was just in the return data from the NVR.
> 
> > Pulling that field, when it is not present would error.
> 
> We schould default it to None when it is not present, can be done in a followup PR.
> 
> > Also, why are you taking a [dict[str,any]](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L34) instead of using your [SearchFile](https://github.com/starkillerOG/reolink_aio/blob/main/reolink_aio/typings.py#L16) typeddict?
> 
> Yea that schould probably indeed be the SearchFile typing, but I just wanted to provide some basics to you of what I had in mind. I did not check if the SearchFile typing is consistant with the NVR response.

On a side note, what is the target of this lib ,10 or 11? if 11 I would like to use StrEnums where possible instead of literal strings, if 10 is a possible target, we can either go with Enums and "fake" an StrEnum or use a backport lib. I kinda wish the one in HA was a separate lib instead of module code, then we could use that because it would already be present. I wasnt sure what your thoughts are on this so I haven used any yet.

**Comment by xannor** [comment]
Thanks for the info, and getting the first part merged. I need some help with deciphering the filenames. The best I can "detect" with the cameras I have is the following:
![image](https://github.com/starkillerOG/reolink_aio/assets/2722473/cabad8eb-e111-49ee-9812-ce45a06c76c7)
I cannot test vehicles because I dont have a camera in a position to accurately detect them. I also know, from the API specs, that there is also face and animal detection types, but I dont have anything that supports that.  

I need a file list that includes the missing types detected. The way I have been doing this is using the reolink client to narrow down a start time for a detection type and then looking in the web client download list for the file listed for that time, downloading and verifying that it is correct, then notice what the client says its detects and what the download filename is.

If I could get a few of these for the mising types, vehicle, face, and animal. then I can complete the "trigger" decoding, until they had more types.

**Comment by starkillerOG** [comment]
@mnpg could you maybe help out here?
Do you have a camera in a position where it can detect vehicles/pets?
Or could you maybe look in the firmware and see how the filenames are constructed for these detection types?

**Comment by mnpg** [comment]
Hi @, hi @xannor ,
I have an e1 outdoor (with md, person and vehicle) and take a look how the search command work to select videos by "tags" (md, person, vehicle and pet) 
I'm busy right now. I contact you when i have infos about this.

**Comment by xannor** [comment]
I decided to try to binwalk some of the firmware as @mnpg suggested. I could get as far as the filenames are constructed in the libStorageFileManager.so and it looks like (from the three different firmwares I checked, though I think I'll double-check my older 511W, non ai one as well) here are all the values supported:
```
FACE DETECTION(%d)
PEOPLE DETECTION(%d)
VEHICLE DETECTION(%d)
ANIMAL DETECTION(%d)
SCHEDULE RECORD(%d)
MD RECORD(%d)
RF RECORD(%d)
```
I am not sure what RF is, MD is motion detection, and SCHEDULE is timer. without access to an, expensive, decompiler or the source, I cannot determine what values it uses or how it encodes them into the three characters I am fairly sure that reflect this in the filename. 
I'll have to wait untill @mnpg  has time to check his cams.

@mpg, as a helper, there is no way, I know of, to search for tags directly, you can only "filter" tags on a day by day basis in the reolink app or desktop client. The web version does not seem to include this feature at all. 

I found out what I have by first finding a day that contained videos for the single tag I was looking for, and determining the start time, and noting what actual "flags" it had, then in the web client, going to that day, opening the downloads and getting the filename from the list with the same start time.

**Comment by mnpg** [comment]
@starkillerOG, hi @xannor : From the firmware of my E1-OutDoor, in my first investigations in the Search command, i ain't seen any "tags" or "flags" depend on the type of what we want to see (md, person, vehicle or cat). I'll be looking further. I'll tell you if i find something.

**Comment by mnpg** [comment]
Hi guys, 
I found something interesting about recordings. In this snapshot I only have MD or PERSON detection. It's seems that the flag is here (details in red)
![image](https://github.com/starkillerOG/reolink_aio/assets/54839700/8f5632b1-783a-4f18-93f3-ad55d63ba6c9)

**Comment by xannor** [comment]
I have seen that as well, I have also had schedule with motion, person, and pet, and pet only, no motion. I think it is because the AI detection can theoretically fire outside of the motion sensitivity. so if you have your regular motion detection set really high, AI can still fire for detecting "motion" that does not trigger the regular motion. Also, and I am not sure why, but as far as I can tell, there are two separate motion detection markers, C08 is bit 4 (int 8) for motion and bit 2 (int 4) for person, 0 for no timer, and bit 4(int 8) for motion. I had thought that maybe the motion was repeated twice for AI detected motion as opposed to alarm detected motion, but I do not have a definite of that. 

Now if I could get the values for vehicle and possible face (though I dont know if they have any cameras that can do face, or if face is a marker for when the camera blurs a face) I would have enough to finish the decoding portion and submit that PR. Then I can work on the integration side, for the media playback and downloading.

**Comment by dbiczo** [comment]
Thanks for the work you guys are putting in. I just checked the recordings on my RLC-520A camera with firmware v3.1.0.1228_22082201 which has AI person & car detection.

The downloaded files are named like:
Mp4Record_2023-05-15_RecM02_20230515_071811_071835_6D28900_13CE8C7.mp4
Mp4Record_2023-05-15_RecM02_20230515_134530_134602_6D28C00_19E7952.mp4
Mp4Record_2023-05-15_RecM02_20230515_092523_092605_6D28D00_214A098.mp4

It looks like the "900" in "6D28900" is car detection,
the "C00" in "6D28C00"  is person detection and
the "D00" in "6D28D00" is car & person detection.

**Comment by mnpg** [comment]
Hi @xannor, with timer only with MD, i have 6D28810. I redo the MD only to verify the value i've sent you.

**Comment by xannor** [comment]
808 is normal MD, 810 is normal timer (I wist I knew what the first 8 meant it seems to always be present.) All I really need now is FACE if there are any cameras that support any form of face detection (it is listed in the API docs.) also I think there is another "alarm" trigger called RF which I believe is and NVR thing but i am not sure.

**Comment by dbiczo** [comment]
@xannor, when I hover over the download list in the web interface I get a path like:
Mp4Record/2023-05-18/RecM02_20230515_071811_071835_6D28900_13CE8C7.mp4
while the downloaded file name is:
Mp4Record_2023-05-15_RecM02_20230515_071811_071835_6D28900_13CE8C7.mp4

**Comment by dbiczo** [comment]
For an old RLC-520 Camera with firmware v3.0.0.136_20121112 which just has motion detection the path on the web interface and downloaded file are both named like:
Rec_20230518_061021_541_M.mp4

I could get the SD cards out of the cameras and get the file paths if you need but it is a bit of pain to do.

**Comment by starkillerOG** [comment]
@xannor thank you very much for all your work on this.
Once you have a stable version working, I would love to test it out and try to get it pulled into the build in integration.

**Comment by xannor** [comment]
I have a semi-stable version working. I am running into a problem with long videos where HA will timeout the pass through connection. Otherwise it seems to work on a basic level.

I have PR #28 to add the lib support I need, once that is in (and an update pushed to HA) I should be able to release the initial version of the media source experiment.

**Comment by starkillerOG** [comment]
@xannor the code has been merged, will release a new lib version in a few days.
I want to also include the firmware check suggested here: https://github.com/starkillerOG/reolink_aio/issues/30 before I push a new version.

**Comment by mnpg** [comment]
Hi @starkillerOG, i'm very pleased if you mention @xannor  and myself in the Official Reolink integration for the work or help we have done to enhance the reolink-api for all the HA reolinkers community. As @xannor, i want to improve it  and share my experience to make this integration much better.
Regards

**Comment by starkillerOG** [comment]
@mnpg I just added you to the reolink-aio 0.6.0 release notes: https://github.com/starkillerOG/reolink_aio/releases/tag/0.6.0
Sorry I did not add you initialy.

In the HomeAssistant documentation itself code-owners/contributors are not mentioned, I myself are also not mentioned in the docs: https://www.home-assistant.io/integrations/reolink/

**Comment by starkillerOG** [comment]
If you find some other place where I schould mention you and other contributors, please let me know and I would be happy to add you.

**Comment by mnpg** [comment]
I think you can mention on the official reolink webpage: https://www.home-assistant.io/integrations/reolink/

**Comment by starkillerOG** [comment]
@mnpg searching through more than 2000 integrations, in none of them the contributors are listed on the docs page, so that is probably agains the HomeAssistant rules (not set by me).

I could however add you to the Acknowledgements section which shows up in this reolink-aio github library https://github.com/starkillerOG/reolink_aio#acknowledgment and on the pypi website https://pypi.org/project/reolink-aio/

Would you like that?

**Comment by mnpg** [comment]
Yes, i agree with your purposes. Thank you.

**Comment by starkillerOG** [comment]
@mnpg I have added you to the readme.

**Comment by hazenjaqdx3** [comment]
Sorry I am very dense and I have not completely understood the meaning of this media platform.
This feature means that I can plug a micro sd card into the camera and it will record to it and I can then watch this recording in HA with your integration? 
The HA doesnt mention this at all :/

**Comment by xannor** [comment]
@hazenjaqdx3 HA doesn't mention it because it is not something that many integrations have implemented, which is part of why it will take a bit to get it working well. 

The end goal is to add [Media Source](https://www.home-assistant.io/integrations/media_source/) support to the integration so any recordings on an ReoLink device (whether IP Camera or NVR) would be available for playback from the HA interface.

Unfortunately there is not a lot of documentation or other integration support around this component so a lot of trial an error is needed to get it right. That is why I am working with an "experimental" component to try to get it to a usable state for more general testing and acceptance. With the eventual goal of getting it in as part of the supported integration.

Hopefully this helps explain things for you.

**Comment by hazenjaqdx3** [comment]
okay and the recordings/media source could for sure be an SD Card?

**Comment by xannor** [comment]
The download/playback is a function of the camera, so as long as the camera can see its storage (and it is a camera the integration can talk to) it should be available in HA, when all is said and done.

**Comment by xannor** [comment]
I made some more progress with the component, I solved the timeout on the HA side with the view, however, I have run into another issue. Due to a design choice in the lib, the aiohttp library, passes the session timeout through to the request as a global timeout. This means that the connection can only run from the initial request until the timeout, regardless of activity. This choice is very annoying, and has caused some vitriol in their issue tracker. The solution is to disable the timeout before the request, so I am looking into how to best do that for a download. This could be why a lot of places use a manual async with TimeoutContext to wrap around their requests instead of the session timeout.

**Comment by xannor** [comment]
Got a new pull submitted, hopefully this would be the last of the needed changes to the API. once this is pushed into HA I should be able to release the HACS addon version. I am also going to see if I can extract the reolink component from the HA install into a "fork" and work on moving the media_source component over for an override style component and future potential pull requests against the component.

**Comment by starkillerOG** [comment]
@xannor I see you indeed figured out how to overwrite the timeout for a specific request.
See my comments on the PR.

Thank you for the greath work on this!

Once you have a more or less stable version I would love to try it out and see what we need to get it into the build-in integration.

**Comment by starkillerOG** [comment]
I have published reolink-aio 0.7.2 for you, so you can continue your work.
Do you already have a branch of HomeAssistant which I can test?

**Comment by xannor** [comment]
Right now I have it as a separate "extension" that depends on the built in component. This was so I could hash out how to implement the platforms, without breaking the integration. The code I have worked on is located [here](https://github.com/xannor/ha_reolink_extras/tree/dev) but it currently requires either the published component to support the updated library, or some manual tweaking of the installed environment to "override" the lib with a development version. I am looking into if I can extract the reolink component into a separate repo to isolate it and work on just the changes it needs to start integrating some of the parts.

**Comment by xannor** [comment]
Sorry for the long hiatus, needed some time to deal with personal issues. 

I am going to fork your fork of HA Core and start building out the needed parts to make this work in the integration. My only stumbling block right now is how to "expose" the videos. Because the search feature is designed around calendar dates, there is not a clean nor practical way to implement it in the Media Browser without some heavy caching or hard use of the API.

This is because the Media Browser expects a root/tree structure, and though I can fake one, it would mean having to "walk" the camera's entire file list every browse step, or cache this and rely on the motion sensors to query updates (this is what I have tried so far. )

A more "natural" way, for the API anyway, would be to implement via calendar. This would allow a search pattern the API expects, and a way to filter because a calendar entity could be created per camera, and the videos would be tagged events. The only real issue with this is that it is not an expected pattern to access videos.

For thumbnails, my thoughts are to make it a config option, off by default, to use the media playback of the camera and HA video lib to grab a snap of a file at a certain time stamp (say 15 secs) then send that to the browser with heavy cache tags, and rely on the browser to hold the image. This could have load issues on cameras if multiple users hit the same camera frequently but have a non-caching client. I feel thumbnails are needed as the info I get from the videos is very limited but I understand the limits they impose on the hardware.

What are your thoughts on this?

**Comment by starkillerOG** [comment]
@xannor I understand, thank you for the update!

I would personally not use the calander for the videos, since to me that seems not to be the intended way in HomeAssistant.
Although I get that it could work very nicely, it feels a bit as a workaround and therefore probably will not be accepted by the core dev team.
They will probably give as feedback that it schould be in the media_browser and if functionality is missing in the media_browser that schould be adressed in core instead of hacking around it by "mis"-using the calander entities.
But of course I am not sure.

I would probably make a "folder"-structure in the media browser to have on the top level the diffrent camera's, then for each camera have a folder "time", "motion", 'people', 'vehicle', 'pet', 'visitor' etc. for the diffrent kinds of detections.
And then in each of those folders have the video files with the name starting with the data like 2023-09-22_14-27-31 (YYYY-MM-DD_HH-MM-SS).

Furthermore to get it into HomeAssistant core, make the PR as small as possible and use the bear minimum of code and functionallity. Start with just adding the basics even if that feels like having a incomplete solution. Then get that merged and make the next PR that will build on top of that to extend the features.
This will make the review process a lot faster and easier and makes it much more likely the PR will get accepted.

Therefore forget about the thumbnails for now, do not include that in the intial PR, we can think about adding those once the basic functionallity is merged.

**Comment by starkillerOG** [comment]
Also in any case do not add media_browser and calendar entities in 1 PR, decide on which you would like to implement first (I would suggest media browser) and then only add that one for the initial PR. The other could always be added in a follow up PR.

**Comment by xannor** [comment]
The problem is you must search using a date, and that is the only filter the API allows. Also the only reliable date to search with is the current date, as to large of a search range (much beyond a month) will just return no results or an error even if there should be some. The only way I have found to find the start date is to "walk" the search backwards a month at a time until there are no status results, or an error. However this is wholly dependent on the size of the memory card and the frequency/length of recording, plus since the cameras will roll off older recordings, once I even find the start I cannot rely on it being present for long. This is why I looked at the calendar since it is the only way to have a window that the API expects. This is either because or why both the web interface and app give you a month calendar to navigate with. The "filters" they provide are just filters applied after the search is run against the encoded filenames returned.

As far as the changes I planned on only implementing what I need for the video files to be accessible to start. Thumbnails was just pre-planning ahead.

**Comment by starkillerOG** [comment]
@xannor what I propose:
The first time a user opens the media browser:
1) request a `await host.request_vod_files(status_only=True, channel=0, start=datetime.now() - timedelta(days=365), end=datetime.now(), stream="sub")`
For me (using a NVR) this returns in 0.35 seconds even if searching for a year back.
However a NVR does store its files diffrently than individual cams and I have 24/7 recording not event based.
This will give you the days on which recording files are present.
Cach this info in 1 variable.
2) Loop from datetime.now() untill the last day at which recordings are present from step 1 in chuncks of lets say 1 month (we can optimize). And cache 1 big list of `typings.VOD_file`, since this is only some text it schould not take much data.

A second time a user opens the media browser:
1) run the same `await host.request_vod_files(status_only=True, channel=0, start=datetime.now() - timedelta(days=365), end=datetime.now(), stream="sub")`
And compare this against the cached previous result.
2) from the cached list of `typings.VOD_file` remove all files on the days not indicated as having recordings anymore + throw away the last day and most recent day.
3) request the last day + loop from datetime.now() untill the previous most recent day.

In that way I think we can make it pretty fast for people regularly using it.


Does running  `await host.request_vod_files(status_only=True, channel=0, start=datetime.now() - timedelta(days=365), end=datetime.now(), stream="sub")` work for your cameras?

**Comment by xannor** [comment]
I will have to test that, but I am pretty sure, with the cameras I have, if there are "too many" recordings in that range it would either return nothing, or an error. The NVR probably handles the load better. 

One good thing, is it does not have to be a full query, it can be a status only, as that just returns the status tables, which include a flag for which days have videos for a given month in the range. (which is how the web and app work, and why the show month to month.)

**Comment by starkillerOG** [comment]
@xannor yes that is the `status_only=True` part of the command, therefore I have hope that it will also work on your cameras to querry all recodings (past year) in one go but then only get the status not the full query.

**Comment by xannor** [comment]
No I think the load issue is that the cameras do not maintain a live DB but instead do a file scan, (I tried diffing through the actual firmware, but cannot verify without source.) And I think if the file scan gets too big it bails, which given the limited memory on the cameras and no swap, makes sense.

What I have done in my tests, is walk backwards month to month (as that seemed to always work) until I got an error or an empty table, then consider that the "start", however it is not fast, not incredibly slow, only about a couple of seconds, depending on the amount of recordings on the camera, and to work around that I would cache the results, and whenever motion was detected use that as a trigger to re-scan the lowest month I found to see if the range changed. This however would get tripped up by a day change or video rolling off, and the timer even does not seem to fire an onvif event so it would get missed if setup on the camera.

I might just leave it to, you only get the current month or two as a safe bet, due to limitations of the media browser and API. This might be why no other camera integration really provides media playback support.

**Comment by starkillerOG** [comment]
I would not trigger on ONVIF motion event, that seems like a bad idea since we could be having a lot of demanding requests to the camera on each motion.

As a first PR I think it would indeed make perfect sense to just provide the last 2 months, I think not many people will actually be using more than that anyway.

How much months of recodings do you have?

**Comment by xannor** [comment]
It varies from camera to camera, I have an older 511W that has no smart AI and a large card, but detects so much motion due to clouds and wind that is keeps barely a month and has many hundreds of recordings on it. But my driveway cam, which only gets a few recordings a day can go back to almost the beginning of the year. LOL, just checked my front porch cam and it goes back to 2021 which I think is when I installed it, yep, the first few videos are from when I was testing in before installation.

**Comment by xannor** [comment]
Here is the approach I am going to take:

I will modify the camera entity, adding the search functionality, download, and playback capability depending on the existing configuration. This will also include adding the appropriate view for the downloads, Also I can add here a "lock" to prevent the attempt to download more than one video at a time since that is a documented API limitation and the cameras do not respond kindly to abuse of it.

I will add a media browser platform entry that will just do a simple dump of a fixed range and some basic interpretation of the filenames, so the recordings can be viewed.

I will add an option to the config to specify a max months, with a default of 1, for the current month only. 

...
I did just think of an interesting modification for the media browser in core. It would be nice if there was a response you could provide that said there was "more" so a concept of infinite scroll or load more could be implemented. This would handle our needs since it would be going from now back to the start of the recording roll.

**Comment by xannor** [comment]
I got the pull started, leaving it draft as I work on the code parts. I see the system pinged you but for others reference here is the PR

https://github.com/home-assistant/core/pull/100764

**Comment by starkillerOG** [comment]
Great!, I indeed got notified by the bots.
Once it is ready for a first test and review, please give me a ping.
Or of course if you have questions.

**Comment by xannor** [comment]
I was just skimming through the API docs, and I noticed that the NVR has a separate download command (NvrDownload), which seems to produce filenames, like the search does for the cameras, but not actually download...? We may need to implement this in order to have a more complete picture of the search results for the NVR. Not sure why it does things differently, but maybe it a throwback to home some multi-monitor apps or appliances work. I'll try to whip up a PR to call this, maybe even integrate it into the regular search API, testing will be difficult though.

**Comment by xannor** [comment]
If you could provide a debug log here of the ability request and a search request for your NVR it would be helpful. I think there are a couple of the abilities we need to key off of for NVRs to properly handle things.

**Comment by starkillerOG** [comment]
@xannor I made some small adjustments to the library, mainly adjusting the VOD filename to UTC timezone for the NVR:
https://github.com/starkillerOG/reolink_aio/commit/a62a285e683a6c5f9d2a9f382947506f265aa782

Now the following script gives the correct URLs that I can use in VLC to view past recordings on my NVR:
```
from reolink_aio import api
from datetime import datetime, timedelta
host = api.Host(host=IP, username=username, password=password)
await host.get_host_data()
vod_files = await host.request_vod_files(channel=0, start=datetime.now() - timedelta(hours=10), end=datetime.now(), stream="sub")
for file in vod_files[1]:
    print(file)
    print(await host.get_vod_source(channel=0, filename=file.file_name, stream="sub"))
```

**Comment by xannor** [comment]
I didnt have a way to test that since I dont have an NVR, but if it works it means the NVR sticks to UTC time instead of "local" time. That code was really just a "fallback" guess anyway. If this does work with no errors, I will revert my ignore NVR commit and we can continue from there.

**Comment by xannor** [comment]
A few observations/suggestions for your media_source.

1) I don't think it is a good idea to find a camera by assuming its entity id will start with "camera."
I feel it would be smarter to import the Camera class from the camera (which should be safe) and instance checking instead. Since you are pulling entities related to a config, that you created, you should not get false positives.

2) I also do not think it is a good idea to "assume" all your data keys will always be config IDs, even though you have control over it, you may need that data space for something else in the future and could cause unforeseen errors.
I think self.hass.config_entries.async_entries(DOMAIN) would be safer.

3) HD cameras actually have 3 streams, main, sub, and ext. sub and ext are equivalent to main/sub in non HD versions. I think one of the abilities notifies this. (this is ability "live" 1 is main/sub/ext 2 is main/sub)

4) The abilities are important, and if you want to use the RTMP playback I think it would be best to first check that the recReplay for the channel is set to 1, this same would go for adding download support in the future, (recDownload)

5) You don't need to actually request all the days, status will give you the month regardless of how many days, if you want to grab two months, just pass the first day of the current month and the last day of the prior. This will be less work on the camera side. (you could even be "sneaky" and use this to walk back a few months with multiple requests or passing multiple actions, each with two separate start end days.)

6) Some cameras use / in their filenames, so you need to either encode it in the url you generate first, or modify your split code. That is why in what I built I used a chain of partition instead of a single split, so I could handle an arbitrary number of / in the identifier.

**Comment by xannor** [comment]
Also I got a new dev environ up to test this, and it does not work for HD cameras, most likely due to the / in the filenames.

```
2023-11-04 14:52:34.015 ERROR (stream_worker) [homeassistant.components.stream] Error from stream worker: Error opening stream (ERRORTYPE_1, Operation not permitted) rtmp://192.168.3.168:1935/vod/Mp4Record?channel=0&stream=1&user=****&password=****
```

usually there is a / and the filename after the Mp4Record

**Comment by xannor** [comment]
Once I added a quick hack to get it to work, it does play, but RTMP playback leaves much to be desired. Most notably the fact that once the stream ends, the player treats it as a broken stream, which causes the player to re-request the stream which starts playback over in a weird matrix glitch fashion.

**Comment by starkillerOG** [comment]
Thanks for the testing and feedback, my biggest concern is the "/", my first thought would be to replace it with "\" as seperator, but I can also chose a different char like "%" or "|", what would you think is the safest char?

Have you tried downloading or streaming a VOD with the ext resolution?
I have not yet seen it, so that is why I did not add it, but I definitely could if it works.

Adding ability checks indeed seems like a good idea, I will do that.

I will also adjust the request of only the first day instead of the full month.

Yea I will first get this PR in, then (hopfully with your help) we can get the download support in for the standalone cameras.

**Comment by xannor** [comment]
I would recommend using a + if your want to replace /, + is already treated by browsers as a space and a space is not part of a valid filename for the camera, which is why they encode it as a %20 on the api side. I don't know how browser would treat a % in the path portion of a url so I would recommend against using %.

**Comment by starkillerOG** [comment]
Sure will make it a +

**Comment by starkillerOG** [comment]
I have optimized the search for VOD files as you suggested by only checking the first 5 minutes of each month.
I did it in the upstream lib instead of the HA code, because that made more sense to me: https://github.com/starkillerOG/reolink_aio/commit/2939e156ee55538e12ae527ce031878e9123d01a

Already tested and it works nicely and is faster :)

**Comment by xannor** [comment]
actually you can do it even simpler, if you have 2+ months, just do from 5 min prior the 1st (the last day of the prior month) to 5 min into the first. This would yield two months per "request", though this would be a micro optimization.

**Comment by xannor** [comment]
The only thing im not sure about is how the cameras would handle an OOB request, empty/error the invalid ones or kill the whole transaction. I never tested multiple search requests for that issue, only singles. when you have the changes in the PR (mostly the new api lib version) let me know and Ill see what happends if I hammer my 511W as it is my oldest camera, wifi, and somewhat of a snowflake.

**Comment by starkillerOG** [comment]
Thanks for the suggestion, I optimized it a bit further as you sugested: https://github.com/starkillerOG/reolink_aio/commit/8e6b02e0ae0bf652254af7c6ca246f4620e89fc6

**Comment by starkillerOG** [comment]
@xannor For the NVR if I go back 5 month in the search I get back this response:
```
[
   {
      "cmd" : "Search",
      "code" : 0,
      "value" : {
         "SearchResult" : {
            "Status" : [
               {
                  "mon" : 10,
                  "table" : "0000000000000000000000001111111",
                  "year" : 2023
               },
               {
                  "mon" : 11,
                  "table" : "111100000000000000000000000000",
                  "year" : 2023
               }
            ],
            "channel" : 0
         }
      }
   },
   {
      "cmd" : "Search",
      "code" : 0,
      "value" : {
         "SearchResult" : {
            "channel" : 0
         }
      }
   },
   {
      "cmd" : "Search",
      "code" : 0,
      "value" : {
         "SearchResult" : {
            "channel" : 0
         }
      }
   }
]
```

So at least the NVR handles this nicely, the API can now also parse such responses correctly

**Comment by xannor** [comment]
I know from testing, some of my older cameras would just error the search request if the range was too large, or if you went  back too far, but that was only with a single search. I never thought to send more than one search to a camera, so I dont know how it would handle it. Hopefully it would only error the individual search but honor others. Will have to test this once the changes are in place.

**Comment by starkillerOG** [comment]
> I don't think it is a good idea to find a camera by assuming its entity id will start with "camera."
> I feel it would be smarter to import the Camera class from the camera [component] (which should be safe) and instance checking instead. Since you are pulling entities related to a config, that you created, you should not get false positives.

I just looked at this, and made it much cleaner now, I am now checking: `entity.domain != CAM_DOMAIN` with CAM_DOMAIN imported from the camera platform.

Thanks for the sugestion!

**Comment by starkillerOG** [comment]
> I also do not think it is a good idea to "assume" all your data keys will always be config IDs, even though you have control over it, you may need that data space for something else in the future and could cause unforeseen errors.
> I think self.hass.config_entries.async_entries(DOMAIN) would be safer.

Thanks for the suggestion, I implemented it and it indeed looks safer now.

**Comment by starkillerOG** [comment]
@xannor I have processed all your feedback, thanks!
The new PR is here: https://github.com/home-assistant/core/pull/103407

Could you maybe test if that new code works with your cams (especially now the seperator is changed to a +)?

**Comment by xannor** [comment]
Yes the new version works, had to manually fix the dev env due to someone else's commit (looks like their making updates to the http component) but I could load the videos from the HD camera without touching your code.

**Comment by xannor** [comment]
However, my 511W threw a too many values to unpack error when selecting a day, after picking res. Investigating...

**Comment by xannor** [comment]
Created pull #48 to fix the issue I discovered.

**Comment by starkillerOG** [comment]
That 511W seems like an odd camera...
Thank you for testing and debugging!

I don't think we need to hold the media_source PR, if it gets merged it will only be included in HA 2023.12.0 at the earliest, by then I will already have a new version bump of reolink-aio (normaly only takes 1 or 2 days for a version push).

**Comment by starkillerOG** [comment]
@xannor I now finished the tests for the PR and added the docs update, it is now ready for a review.
Would you be so kind to give it a review: https://github.com/home-assistant/core/pull/103407

**Comment by xannor** [comment]
I never have done a review though GH before, but as I dont have write access, not sure how much help it will provide.

**Comment by starkillerOG** [comment]
@xannor as you can see in the PR, the media source PR has been merged and will therefore be included in the HA 2023.12 release :)

If you want we can now build upon this first implementation to handle viewing better for standalone cams that can use the download feature.

**Comment by xannor** [comment]
My thoughts are to detect and provide full size downloads where supported, and possibly "abuse" the navigation to support month to month scrolling (i.e. first entry being the previous month, and the last being the next month, when available). This should handle most of the needs for playback support. The only exception would be the rare circumstance of a camera that only records once every two months, and that I feel is too much an outlier.

**Comment by starkillerOG** [comment]
Oh yes, month to month scrolling sounds very nice, did not think of that yet.
Would be a greath addition :)

Full size downloads would indeed also be a nice improvement.

**Comment by xannor** [comment]
Working on the full dl part now. will possibly have a test for you to try (to make sure NVRs dont get tripped up by it) soon. 

my thoughts on the month to month is to take advantage of the status return, and encode the next/current month into the identity (or use some sort of session/temp, I have what question in the HA discord). This way in a given month page we should know if the prior month we came from so we only need to do a status for the beginning or end of the month we want. This would also work since the starting month is current and only the prior month would need to be included (which we can do easily.)

**Comment by starkillerOG** [comment]
Yea I think we can just make 2 extra `BrowseMediaSource` items, one for the next and one for the previous month. (make them conditionally if we are already at the current month and if the previous month actually has recordings).
In the `identifier` of those two items we can just encode to wich new month we schould go.
I don't think it should be too difficult to implement.

**Comment by xannor** [comment]
It would also be good to keep track of which direction we came from, so we only have to query in one direction on the camera, which is why I had the question about session/temp (in discord), otherwise could encode it in the identifier, but wanted to avoid polluting it if possible.

**Comment by xannor** [comment]
Got the basics for DL support here https://github.com/xannor/ha-core/tree/reolink_vod_dl however, having an issue where the download is aborted about 30 seconds in, dont know if it is because of aio timeouts or something in the api side. Just get a connection closed notice. Give it a test and see if it causes any problems with NVRs though.

**Comment by starkillerOG** [comment]
@xannor sorry for taking so long, I just tested your reolink_vod_dl branch with my NVR but it does not work.
`host.api.api_version("recDownload", channel)` is also 1 for all channels of the NVR, (because it does support downloading, but in a diffrent way).
So the NVR now also attempts to do a download and fails.

So we will need to check using `host.api.is_nvr` and make sure the NVR is still using the streaming instead of downloading.
Later we could also try and see if we can get the downloading working on the NVR (but I think we would need a sequence of two commands to get that working).

**Comment by starkillerOG** [comment]
This commit implemented download support for the NVR: https://github.com/starkillerOG/reolink_aio/commit/27e512495b8214ad3c7887c11b002c84da61b11c
It is tested and works (with some small adjustments to your branch).

However I would not recommend it for an NVR since it takes a really long time to start a playback (NvrDownload takes something like 20 seconds before completing).
Moreover it is easy to upset the NVR, when requesting the same download for instance to fast after eachother or opening another download while the first one was not done closing yet (clicking fast in HA) causes the NVR to return a "busy" error. Then you have to wait something like an hour or reboot the NVR to get it out of the "busy" state....
And the Ext stream is not supported for NVRs.

I will make a PR for your branch to suggest the nessesary changes.

Maybe we can disable download for NVRs at first and in a seperate PR once this is merged make a new config flow option to let the user chose between streaming/downloading for playback (the NVR will have streaming as default and other cams downloading as default).
Or we could make options in the browsing navigation (like at the resolution page, make Streaming/Downloading options...)

**Comment by starkillerOG** [comment]
See my PR to your branch here: https://github.com/xannor/ha-core/pull/1

**Comment by xannor** [comment]
I think for the time being, just not doing the download on the NVR is the better approach, as there is little advantage in that instance.

I still feel that an NVR is fundementally the same approach as here and it would be better to manage/use that interface and try to use the limited API and have HA do it.

I mostly want this for those who do not have/want an NVR and want to have some of the functionality in concept.

**Comment by xannor** [comment]
I looked at your PR. the live ability should be 1 if main/sub/ext is supported and 2 if only main/sub according to the API docs. you said the NVR does not support ext, does it report a 1 for that value, because I though I was explicitly checking for 1 in hasExt.

**Comment by starkillerOG** [comment]
Yes the NVR was reporting 1 for the "live" ability, but it only has main/sub playback, not ext.
It does support live realtime streaming on RTMP or FLV, (not RTSP).
The download did not work when Ext was used as stream.

**Comment by xannor** [comment]
its nice to see the NVR does not follow the documented API correctly.. :) I'll add the NVR check to that as well.

**Comment by xannor** [comment]
I did some small edits to disable NVR DL and fully disabled ext playback, for now. I still need to investigate the DL issues I have seen, but could you check that the NVR is working as it should now?

**Comment by xannor** [comment]
> Yes h265 encoding can be a problem to play in some browsers, that is also why the RTSP stream from Reolink does not yet support h265 encoding I think.

I beleive the RTSP does do h265, as that is the only way to pull the full stream, RTMP does not support h265.

> 
> I would just do only sub stream for cameras that have h265 encoding, just like what is done for the NVR.
> 

Yeah, might have to but I dont like it. Would be nice if HA had a download component, for downloading from the HA, and not just downloading to it.

> > I changed the player to pass the url to the camera though
> 
> I don't quite understand this...

Just like with the screen capture API command, the download command can be passed to the camera as a GET style url request. I manually built that in the code to test, and it does work, except for the afformentioned h265 problem.

> 
> > The alternative is to pass the url through (with token) but that both potentially exposes security info, and would not work with clients that do not have direct access to the camera themselves.
> 
> I also don't understand this, could you give an example of the URL you are talking about? (you can replace token/password with xxxx)

When passing the download command as a url, you have to supply the token parameter as well. If we send this url to the client, it means internal security information (normally the token is not exposed anywhere) would be sent over the web socket. Even though this should be safe, I dont like the thought of doing that, if it can be avoided.

The reason I tried this approach is because all view requests in HA, by default use the default timeout of 30 seconds for the total life of the session connection. Since proxing the file through HA means two connections, one on the camera side the integration can control, and one on the HA side that core controlls, if the file takes longer than 30 seconds to transfer, it gets killed by the timeout. Most players will only buffer about 10-15 seconds and pause thedownload, so almost all transfer of files larger than a 15 second video will fail.

Passing the url through works around this because now HA is not involved with the transfer, however, this will not work in situations where the client must access content though HA (such as remote connections) as the url would not be properly resolvable to the client.

In addition if the connection is going over SSL, and a self-signed cert is used. then the player will not work anyways because the browser would reject the cert, proxying through HA also worked around this issue.

Also, passing the url like that to potentially a remote client means that url would now be re-transmitted over the internet when the client attempts to connect.

Both of these situations make me not want to try to hand the url off to the client as it would lead to a bad experience that a regular user may not realize is not an error in the integration. Thus needless bug tickets.

**Comment by xannor** [comment]
From the absence in the write-up, I'm guessing the media support did not make the cut for this release? Updating right now so I will know for sure in a few minutes.

**Comment by xannor** [comment]
Yep, i see it, surprised there was no mention in the write-up.

**Comment by starkillerOG** [comment]
We could localize if you want.
You can probably get the global HA settings for time format configured.
Then we could use that to dynamically change the title style to 24 vs 12 PM...

**Comment by xannor** [comment]
is there a way to see the user preferences? Going that route I would prefer to use user settings instead of system, if possible.

I have tried to dig through other integrations to see how they handle that sort of thing, but mostly when localization has been involved, they only talk about states and other things done through the lang files, not regular output.

**Comment by xannor** [comment]
I was re-reading the API and I see a "new" command called playback that I didnt see before. It is structured exactly like download but does not list the same limitations. I am going to test with it and expand the vod_source function to handle generate one of the three possible urls.

Also in vod source I see you check for rtmp and enable it, this should be the responsibility of the integration to en/force not as a side-effect of the library (generally side-effects are a bad idea). I am going to remove this from my changes.

**Comment by xannor** [comment]
Ok got a little farther, and feel a little better about it. I have a new PR to this lib for the changes I need (and I updated my branch of core). If you could test what I have with your NVR to make sure I didnt break anything, this would be a good start for the second round of media support.

**Comment by starkillerOG** [comment]
I will try looking at it soon, am a bit swamped with work at the moment tho...

Seems like a nice way to go forward.

**Comment by starkillerOG** [comment]
@xannor I just merged the PR (in a seperate comment), I will release a new reolink_aio version shortly.

**Comment by xannor** [comment]
Ahh, I see where you thought things didnt work. I was not enforcing the RTMP/DL limit in API but in the component, I figure the API should be closer to "dumb" and the calling code should enforce any device specific rules, except where the API docs state such a rule, or a work around is necessary.

**Comment by starkillerOG** [comment]
I like to put as much logic in the library and make the HomeAssistant code as "dumb" as possible.
Because changes in the library are easy and a HA version bump will be done within a day.
However changing logic in HA can take weeks because of waiting on reviews.....

**Comment by xannor** [comment]
Yeah, I see where your coming from now. Its just an old habit of mine, to write to the as close to the API and UI as possible and bridge the difference.

**Comment by starkillerOG** [comment]
@xannor version bump PR in HA is here: https://github.com/home-assistant/core/pull/106747
I needed to add a small bit of logic to restore the default behaviour of RTMP for seperate cams and FLV for NVRs.

**Comment by starkillerOG** [comment]
@xannor could you check this issue: https://github.com/home-assistant/core/issues/110939
Is there something wrong with the playback URL?
I don't have standalone cams with SD cards, so can not easily check myself.

**Comment by starkillerOG** [comment]
With this latest PR, a Proxy is added such that playback inside HomeAssistant is now basically always possible: https://github.com/home-assistant/core/pull/133916
Therefore I will close this issue.
